scripts-available/CDB_CartodbfyTable.sql

```diff
@@ -649,3 +649,649 @@ BEGIN
+
+  -- It's perfect (named right, right type, right index)!
+  IF FOUND AND rec.indisprimary AND rec.indisunique AND rec.attnotnull AND rec.atttypid IN (20,21,23) THEN
+    RAISE DEBUG '_CDB_Has_Usable_Primary_ID found good ''%''', keyname;
+    RETURN true;
+    
+  -- It's not suitable (not an integer?, not unique?) to rename it out of the way
+  ELSIF FOUND THEN
+    RAISE DEBUG '_CDB_Has_Usable_Primary_ID found bad ''%'', renaming it', keyname;
+    
+    sql := Format('ALTER TABLE %s RENAME COLUMN %s TO %s', 
+              reloid::text, rec.attname, _CDB_Unique_Column_Name(reloid, keyname));
+    EXECUTE sql;        
+    
+  -- There's no column there named keyname
+  ELSE
+
+    -- Is there another suitable primary key already?
+    SELECT a.attname
+    INTO rec
+    FROM pg_class c 
+    JOIN pg_attribute a ON a.attrelid = c.oid 
+    JOIN pg_type t ON a.atttypid = t.oid
+    LEFT JOIN pg_index i ON c.oid = i.indrelid AND a.attnum = ANY(i.indkey)
+    WHERE c.oid = reloid AND NOT a.attisdropped
+    AND i.indisprimary AND i.indisunique AND a.attnotnull AND a.atttypid IN (20,21,23);
+  
+    -- Yes! Ok, rename it.
+    IF FOUND THEN
+      EXECUTE Format('ALTER TABLE %s RENAME COLUMN %s TO %s', reloid::text, rec.attname, keyname);
+      RAISE DEBUG '_CDB_Has_Usable_Primary_ID found acceptable primary key ''%s'', renaming to ''%''', rec.attname, keyname;
+      RETURN true;
+    ELSE
+      RAISE DEBUG '_CDB_Has_Usable_Primary_ID found no useful column for ''%''', keyname;
+    END IF;
+  
+  END IF;
+
+  -- Remove any unsuitable primary key constraint that is hanging around, 
+  -- because we will be adding one back later
+  SELECT ci.relname AS pkey
+  INTO rec
+  FROM pg_class c 
+  JOIN pg_attribute a ON a.attrelid = c.oid 
+  LEFT JOIN pg_index i ON c.oid = i.indrelid AND a.attnum = ANY(i.indkey)
+  JOIN pg_class ci ON i.indexrelid = ci.oid
+  WHERE c.oid = reloid AND NOT a.attisdropped
+  AND a.attname != keyname
+  AND i.indisprimary AND a.atttypid NOT IN (20,21,23);
+  
+  IF FOUND THEN
+    EXECUTE Format('ALTER TABLE %s DROP CONSTRAINT IF EXISTS %s', reloid::text, rec.pkey);
+    RAISE DEBUG '_CDB_Has_Usable_Primary_ID dropping unused primary key ''%''', rec.pkey;
+  END IF;
+    
+  -- Didn't fine re-usable key, so return FALSE
+  RETURN false;
+
+END;
+$$ LANGUAGE 'plpgsql';
+
+
+CREATE OR REPLACE FUNCTION _CDB_Unique_Relation_Name(schemaname TEXT, relationname TEXT)
+RETURNS TEXT
+AS $$
+DECLARE
+  rec RECORD;
+  i INTEGER;
+  newrelname TEXT;
+BEGIN
+
+  i := 0;
+  newrelname := relationname;
+  LOOP
+
+    SELECT c.relname, n.nspname
+    INTO rec
+    FROM pg_class c
+    JOIN pg_namespace n ON c.relnamespace = n.oid
+    WHERE c.relname = newrelname
+    AND n.nspname = schemaname;
+  
+    IF NOT FOUND THEN
+      RETURN newrelname;
+    END IF;
+    
+    i := i + 1;
+    newrelname := relationname || '_' || i;
+  
+    IF i > 100 THEN
+      RAISE EXCEPTION '_CDB_Unique_Relation_Name looping too far';
```

Comment:
I guess we would have to add all the new exceptions to the calling code (importer) so they can be managed, wouldn't we?